server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceSimulationTests.java

```diff
@@ -0,0 +1,289 @@
+import org.elasticsearch.cluster.DiskUsage;
+import org.elasticsearch.cluster.ESAllocationTestCase;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.routing.IndexRoutingTable;
+import org.elasticsearch.cluster.routing.RoutingTable;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.ShardRoutingState;
+import org.elasticsearch.cluster.routing.TestShardRouting;
+import org.elasticsearch.cluster.routing.allocation.DataTier;
+import org.elasticsearch.cluster.routing.allocation.WriteLoadForecaster;
+import org.elasticsearch.cluster.routing.allocation.decider.ClusterRebalanceAllocationDecider;
+import org.elasticsearch.cluster.routing.allocation.decider.ConcurrentRebalanceAllocationDecider;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.xcontent.json.JsonXContent;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+public class DesiredBalanceSimulationTests extends ESAllocationTestCase {
+
+    public void testDesiredBalanceFromDesiredBalanceOutput() throws IOException {
+
+        var data = parse();
+
+        logger.info("Initial stats {}", stats(data));
+
+        Settings settings = Settings.builder()
+            // .put(
+            // ClusterModule.SHARDS_ALLOCATOR_TYPE_SETTING.getKey(),
+            // ClusterModule.DESIRED_BALANCE_ALLOCATOR
+            // )
+            .put(
+                ClusterRebalanceAllocationDecider.CLUSTER_ROUTING_ALLOCATION_ALLOW_REBALANCE_SETTING.getKey(),
+                ClusterRebalanceAllocationDecider.ClusterRebalanceType.INDICES_PRIMARIES_ACTIVE
+            )
+            .put(ConcurrentRebalanceAllocationDecider.CLUSTER_ROUTING_ALLOCATION_CLUSTER_CONCURRENT_REBALANCE_SETTING.getKey(), 50)
+            .build();
+
+        // var allocator = createShardsAllocator(settings);
+        // var allocation = new RoutingAllocation(
+        // randomAllocationDeciders(settings, ClusterSettings.createBuiltInClusterSettings(settings)),
+        // data.clusterState.mutableRoutingNodes(),
+        // data.clusterState,
+        // data.clusterInfo,
+        // SnapshotShardSizeInfo.EMPTY,
+        // 0L
+        // );
+        // allocator.allocate(allocation, ActionListener.noop());
+        //
+        // logger.info("Same: {}, initializing: {}",
+        // data.clusterState.getRoutingNodes().equals(allocation.routingNodes()),
+        // RoutingNodesHelper.shardsWithState(allocation.routingNodes(), INITIALIZING).size()
+        // );
+
+        var service = createAllocationService(settings, () -> data.clusterInfo);
+        var newClusterState = applyStartedShardsUntilNoChange(data.clusterState, service);
```

Comment:
`data.clusterInfo` needs to be updated after each iteration to reflect actual shard movements